In [4]:
# Data handling and manipulation library
import pandas as pd

# Library for numerical operations in Python
import numpy as np

In [5]:
""" country = pd.read_csv('Tabel_SKS_Diagnose_country.csv', sep=';')
country.head()
country.shape """

" country = pd.read_csv('Tabel_SKS_Diagnose_country.csv', sep=';')\ncountry.head()\ncountry.shape "

In [6]:
region = pd.read_csv('Tabel_SKS_Diagnose_region.csv', sep=';')
region.head()


Aar     SKS                               SKS_Kodetekst Diagnose  \
0  2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi   DI702A   
1  2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi    DM890   
2  2018  BABZ00               Anlæggelse Af Epiduralkateter    DA419   
3  2018  BABZ00               Anlæggelse Af Epiduralkateter    DC153   
4  2018  BABZ00               Anlæggelse Af Epiduralkateter    DC154   

             Region  Unikke_Patienter  Registreringer  
0  1081 Nordjylland                 7            12.0  
1  1081 Nordjylland                 3             5.0  
2  1082 Midtjylland                 3             3.0  
3  1082 Midtjylland                 4             7.0  
4  1082 Midtjylland                 9            11.0

In [7]:
hospital = pd.read_csv('Tabel_SKS_Diagnose_hospital.csv', sep=';')
hospital.head()

Aar     SKS                               SKS_Kodetekst Diagnose  \
0  2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi   DI702A   
1  2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi    DM890   
2  2018  BABZ00               Anlæggelse Af Epiduralkateter    DC153   
3  2018  BABZ00               Anlæggelse Af Epiduralkateter    DC154   
4  2018  BABZ00               Anlæggelse Af Epiduralkateter    DC155   

   Sygehuskode                       Sygehus  Unikke_Patienter  Registreringer  
0         8001  Aalborg Universitetshospital               7.0              12  
1         8001  Aalborg Universitetshospital               3.0               5  
2         6620   Aarhus Universitetshospital               4.0               7  
3         6620   Aarhus Universitetshospital               9.0              11  
4         6620   Aarhus Universitetshospital              11.0              11

In [8]:
#country, 
region, hospital

(         Aar     SKS                               SKS_Kodetekst Diagnose  \
 0       2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi   DI702A   
 1       2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi    DM890   
 2       2018  BABZ00               Anlæggelse Af Epiduralkateter    DA419   
 3       2018  BABZ00               Anlæggelse Af Epiduralkateter    DC153   
 4       2018  BABZ00               Anlæggelse Af Epiduralkateter    DC154   
 ...      ...     ...                                         ...      ...   
 456841  2018  ZZV030           Vurdering Ved Brug Af Spørgeskema    DR429   
 456842  2018  ZZV030           Vurdering Ved Brug Af Spørgeskema    DZ033   
 456843  2018  ZZV030           Vurdering Ved Brug Af Spørgeskema   DZ033A   
 456844  2018  ZZV030           Vurdering Ved Brug Af Spørgeskema    DZ038   
 456845  2018  ZZV030           Vurdering Ved Brug Af Spørgeskema    DZ039   
 
                   Region  Unikke_Patienter  Registreringer  


In [9]:
df = pd.concat([region, hospital], axis=0, ignore_index=True)
df

Aar             SKS                               SKS_Kodetekst  \
0        2018           BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi   
1        2018           BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi   
2        2018          BABZ00               Anlæggelse Af Epiduralkateter   
3        2018          BABZ00               Anlæggelse Af Epiduralkateter   
4        2018          BABZ00               Anlæggelse Af Epiduralkateter   
...       ...             ...                                         ...   
1226137  2018          ZZ0150                            Journaloptagelse   
1226138  2015  Diskretioneret                              Diskretioneret   
1226139  2017  Diskretioneret                              Diskretioneret   
1226140  2018  Diskretioneret                              Diskretioneret   
1226141  2017  Diskretioneret                              Diskretioneret   

        Diagnose            Region  Unikke_Patienter  Registreringer  \
0         DI702A  1081 Nordjylland               7.0            12.0   
1          DM890  1081 Nordjylland               3.0             5.0   
2          DA419  1082 Midtjylland               3.0             3.0   
3          DC153  1082 Midtjylland               4.0             7.0   
4          DC154  1082 Midtjylland               9.0            11.0   
...          ...               ...               ...             ...   
1226137   DZ980C               NaN               3.0             3.0   
1226138    DZ988               NaN               6.0             6.0   
1226139    DZ988               NaN               8.0            49.0   
1226140    DZ988               NaN              17.0            52.0   
1226141      NaN               NaN               3.0             9.0   

         Sygehuskode   Sygehus  
0                NaN       NaN  
1                NaN       NaN  
2                NaN       NaN  
3                NaN       NaN  
4                NaN       NaN  
...              ...       ...  
1226137       8887.0  Slagelse  
1226138       8887.0  Slagelse  
1226139       8887.0  Slagelse  
1226140       8887.0  Slagelse  
1226141       8887.0  Slagelse  

[1226142 rows x 9 columns]

In [10]:
# List of DataFrames
#dfs = [ region, hospital]

kattemad = pd.merge(region, hospital, left_on=["SKS", "Diagnose"], right_on=["SKS", "Diagnose"])


""" combined_df = dfs[0] # hashtag så Jasmin forstår det betyrder at vi combinerer dataframes der starter på 0
for i, df in enumerate(dfs[1:], start=1):
    combined_df = pd.merge(combined_df, df, on=['SKS'], how='outer', suffixes=('', f'_df{i+1}'))
#columns_to_drop = ['brick_nr_df2', 'brick_nr_df3','brick_nr_df4','brick_nr_df5','brick_nr_df6','brick_nr_df7','brick_nr_df8']

# Drop the specified columns
df = combined_df """
kattemad.head()

Aar_x     SKS                             SKS_Kodetekst_x Diagnose  \
0   2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi   DI702A   
1   2018   BABZ0  Anlæggelse Af Kateter Til Regionalanæstesi    DM890   
2   2018  BABZ00               Anlæggelse Af Epiduralkateter    DC153   
3   2018  BABZ00               Anlæggelse Af Epiduralkateter    DC154   
4   2018  BABZ00               Anlæggelse Af Epiduralkateter    DC155   

             Region  Unikke_Patienter_x  Registreringer_x  Aar_y  \
0  1081 Nordjylland                   7              12.0   2018   
1  1081 Nordjylland                   3               5.0   2018   
2  1082 Midtjylland                   4               7.0   2018   
3  1082 Midtjylland                   9              11.0   2018   
4  1082 Midtjylland                  11              11.0   2018   

                              SKS_Kodetekst_y  Sygehuskode  \
0  Anlæggelse Af Kateter Til Regionalanæstesi         8001   
1  Anlæggelse Af Kateter Til Regionalanæstesi         8001   
2               Anlæggelse Af Epiduralkateter         6620   
3               Anlæggelse Af Epiduralkateter         6620   
4               Anlæggelse Af Epiduralkateter         6620   

                        Sygehus  Unikke_Patienter_y  Registreringer_y  
0  Aalborg Universitetshospital                 7.0                12  
1  Aalborg Universitetshospital                 3.0                 5  
2   Aarhus Universitetshospital                 4.0                 7  
3   Aarhus Universitetshospital                 9.0                11  
4   Aarhus Universitetshospital                11.0                11

In [11]:
dosing = pd.read_csv('Dosing.csv', sep=';')
dosing.head()


Substance  Dose Index Diagnosis Group
0  Certolizumab        4800             PSA
1  Certolizumab        4800              RA
2    Infliximab        3000              CD
3    Infliximab        3000             PSA
4    Infliximab        3000              RA

In [12]:
substances = pd.read_csv('substancesOfInterest.csv', sep=';')
substances.head()

ATC5       SKS     Substance Medicinkode
0  L04AB02  BOHJ18A1    Infliximab    ML04AB02
1  L04AB05  BOHJ18A5  Certolizumab    ML04AB05
2  L04AC05  BOHJ18B3   Ustekinumab    ML04AC05

In [13]:
substances, dosing

(      ATC5       SKS     Substance Medicinkode
 0  L04AB02  BOHJ18A1    Infliximab    ML04AB02
 1  L04AB05  BOHJ18A5  Certolizumab    ML04AB05
 2  L04AC05  BOHJ18B3   Ustekinumab    ML04AC05,
       Substance  Dose Index Diagnosis Group
 0  Certolizumab        4800             PSA
 1  Certolizumab        4800              RA
 2    Infliximab        3000              CD
 3    Infliximab        3000             PSA
 4    Infliximab        3000              RA
 5   Ustekinumab           2              CD)

In [14]:
pd.merge(dosing, substances, left_on=["Substance"], right_on=["Substance"])

Substance  Dose Index Diagnosis Group     ATC5       SKS Medicinkode
0  Certolizumab        4800             PSA  L04AB05  BOHJ18A5    ML04AB05
1  Certolizumab        4800              RA  L04AB05  BOHJ18A5    ML04AB05
2    Infliximab        3000              CD  L04AB02  BOHJ18A1    ML04AB02
3    Infliximab        3000             PSA  L04AB02  BOHJ18A1    ML04AB02
4    Infliximab        3000              RA  L04AB02  BOHJ18A1    ML04AB02
5   Ustekinumab           2              CD  L04AC05  BOHJ18B3    ML04AC05

In [15]:
df = pd.concat([region, hospital], axis=0, ignore_index=True)

In [16]:
sales_hospital = pd.read_excel('salgs_data_indication_insight_hospital (1).xlsx')
sales_hospital['PRIS'] = sales_hospital['PRIS'].round(2)
sales_hospital.head()

ImportError: Missing optional dependency 'openpyxl'.  Use pip or conda to install openpyxl.

In [92]:
sales_pharma = pd.read_csv('salgs_data_indication_insight_pharmacy.csv', sep=';')
column_to_drop = sales_pharma.drop("Unnamed: 10", axis=1)
sales_pharma = column_to_drop
sales_pharma.head()

Year Month           Region Account Description Product    Size  \
0    201003.0  Danmark diverse   0009 GROSSISTSALG  Cimzia  2 stk.   
1    201006.0  Danmark diverse   0009 GROSSISTSALG  Cimzia  2 stk.   
2    201007.0  Danmark diverse   0009 GROSSISTSALG  Cimzia  2 stk.   
3    201011.0  Danmark diverse   0009 GROSSISTSALG  Cimzia  2 stk.   
4    201103.0  Danmark diverse   0009 GROSSISTSALG  Cimzia  2 stk.   

   Size Numeric Strength WHO ATC 5 Code  Volume    Value  
0           2.0   200 mg        L04AB05     1.0  8083,55  
1           2.0   200 mg        L04AB05     1.0  8083,55  
2           2.0   200 mg        L04AB05     1.0  8083,55  
3           2.0   200 mg        L04AB05     2.0  16167,1  
4           2.0   200 mg        L04AB05     1.0  8406,89

In [93]:
df = pd.concat([sales_hospital, sales_pharma], axis=0, ignore_index=True)
#pd.concat(sales_hospital, sales_pharma, left_on=["Product", "Size", "Volume", "Value", "Region", "WHO ATC 5 Code"], right_on=["Product", "Size", "Volume", "Value", "Region", "WHO ATC 5 Code"])

In [122]:
hospital_account = pd.read_csv('Hospital_Account.csv', sep=";")
column_to_drop_hospital = hospital_account.drop("Unnamed: 4", axis=1)
hospital_account = column_to_drop_hospital

hospital_account["Hospital Code"] = pd.to_numeric(hospital_account["Hospital Code"])
hospital_account.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 63 entries, 0 to 62
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Hospital Code  62 non-null     float64
 1   Hospital       63 non-null     object 
 2   Account        51 non-null     object 
 3   Account no.    50 non-null     float64
dtypes: float64(2), object(2)
memory usage: 2.1+ KB


In [123]:
hospital_region = pd.read_csv('Hospital_Region.csv', sep=";")
hospital_region.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Hospital       34 non-null     object
 1   Region         34 non-null     object
 2   Hospital Code  34 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 948.0+ bytes


In [135]:
mjau_mjau_mjau = pd.merge(hospital_account, hospital_region, left_on=["Hospital Code", "Hospital"], right_on=["Hospital Code", "Hospital"])
mjau_new = mjau_mjau_mjau.drop("Account", axis=1)
mjau_new

Hospital Code                                Hospital  Account no.  \
0          2000.0             Hospitalerne I Nordsjælland       4060.0   
1          1516.0             Herlev Og Gentofte Hospital       4060.0   
2          1309.0  Bispebjerg Og Frederiksberg Hospitaler       4060.0   
3          1501.0                       Gentofte Hospital       4060.0   
4          1301.0                          Rigshospitalet       4060.0   
5          4001.0                      Bornholms Hospital       4060.0   
6          1330.0             Amager Og Hvidovre Hospital       4060.0   
7          1502.0                       Glostrup Hospital       4060.0   
8          5501.0                     Sydvestjysk Sygehus       4120.0   
9          6007.0          Fredericia Og Kolding Sygehuse       4120.0   
10         4202.0         Ouh Odense Universitetshospital       4120.0   
11         5000.0                   Sygehus Sønderjylland       4120.0   
12         6008.0                           Vejle Sygehus       4120.0   
13         8001.0            Aalborg Universitetshospital       4130.0   
14         6650.0                   Hospitalsenheden Vest       4100.0   
15         6006.0               Regionshospitalet Horsens       4100.0   
16         6620.0             Aarhus Universitetshospital       4100.0   
17         6630.0                     Hospitalsenhed Midt       4100.0   
18         7005.0               Regionshospitalet Randers       4100.0   
19         3502.0                        Nykøbing Falster       4070.0   
20         3501.0                                 Næstved       4070.0   
21         8881.0                                Roskilde       4070.0   
22         8882.0                                    Køge       4070.0   
23         8883.0                                   Fakse       4070.0   
24         8884.0                                  Holbæk       4070.0   
25         8885.0                              Kalundborg       4070.0   
26         8886.0                                 Næstved       4070.0   
27         8887.0                                Slagelse       4070.0   
28         8888.0                                Ringsted       4070.0   
29         8889.0                        Nykøbing Falster       4070.0   
30         8890.0                                 Nakskov       4070.0   
31         8891.0                                  Maribo       4070.0   

         Region  
0   Hovedstaden  
1   Hovedstaden  
2   Hovedstaden  
3   Hovedstaden  
4   Hovedstaden  
5   Hovedstaden  
6   Hovedstaden  
7   Hovedstaden  
8    Syddanmark  
9    Syddanmark  
10   Syddanmark  
11   Syddanmark  
12   Syddanmark  
13  Nordjylland  
14  Midtjylland  
15  Midtjylland  
16  Midtjylland  
17  Midtjylland  
18  Midtjylland  
19     Sjælland  
20     Sjælland  
21     Sjælland  
22     Sjælland  
23     Sjælland  
24     Sjælland  
25     Sjælland  
26     Sjælland  
27     Sjælland  
28     Sjælland  
29     Sjælland  
30     Sjælland  
31     Sjælland

In [94]:
""" merged_data = pd.merge(sales_pharma, sales_hospital, on="Product", how="left")
merged_data """

""" for index, row in sales_pharma.iterrows():
    product_name = row['Product']
    if product_name in company_name_map:
        products_only.at[index, 'company_name'] = company_name_map[product_name]

# Save the updated products_only dataset
products_only.to_csv("products_only_with_companies.csv", index=False) """

#pd.merge(sales_hospital, sales_pharma, left_on=["Product", "Size", "WHO ATC 5 Code", "Strength"], right_on=["Product", "Size", "WHO ATC 5 Code", "Strength"])

' for index, row in sales_pharma.iterrows():\n    product_name = row[\'Product\']\n    if product_name in company_name_map:\n        products_only.at[index, \'company_name\'] = company_name_map[product_name]\n\n# Save the updated products_only dataset\nproducts_only.to_csv("products_only_with_companies.csv", index=False) '

## Data cleaning and preprocessing
In the first part of this notebook, the focus will be on preparing and cleaning the data for subsequent analysis. This is an example dataset, and does not provide a full picture of all the datasets.

#### 1. Get an overview
We will first investigate the dataframe and look at the value frequency of each column.

This can be used for understanding which column provide enough date to be relevant. Or the correct data type.

Understanding which columns need to be dropped.

In [95]:
df

Year Month       Region                       Account Description  \
0        201001.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
1        201001.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
2        201001.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
3        201002.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
4        201002.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
...           ...          ...                                       ...   
59765         NaN          NaN                                       NaN   
59766         NaN          NaN                                       NaN   
59767         NaN          NaN                                       NaN   
59768         NaN          NaN                                       NaN   
59769         NaN          NaN                                       NaN   

        Product     Company              Size  Size Numeric      Strength  \
0        Cimzia  UCB Nordic            2 stk.           2.0        200 mg   
1      Remicade         MSD  1 htgl. a 100 mg         100.0        100 mg   
2       Stelara     Janssen             1 stk           1.0  45 mg/0,5 ml   
3        Cimzia  UCB Nordic            2 stk.           2.0        200 mg   
4      Remicade         MSD  1 htgl. a 100 mg         100.0        100 mg   
...         ...         ...               ...           ...           ...   
59765       NaN         NaN               NaN           NaN           NaN   
59766       NaN         NaN               NaN           NaN           NaN   
59767       NaN         NaN               NaN           NaN           NaN   
59768       NaN         NaN               NaN           NaN           NaN   
59769       NaN         NaN               NaN           NaN           NaN   

      WHO ATC 5 Code  Volume      Value      PRIS  
0            L04AB05    19.0   160820.1   8464.22  
1            L04AB02  2096.0  9236652.8   4406.80  
2            L04AC05    23.0  622716.26  27074.62  
3            L04AB05    11.0   88919.05   8083.55  
4            L04AB02  2036.0  8972244.8   4406.80  
...              ...     ...        ...       ...  
59765            NaN     NaN        NaN       NaN  
59766            NaN     NaN        NaN       NaN  
59767            NaN     NaN        NaN       NaN  
59768            NaN     NaN        NaN       NaN  
59769            NaN     NaN        NaN       NaN  

[59770 rows x 12 columns]

In [96]:
#To get a sense of which features offer meaningful information, we print out the frequency of each
#value for each feature
cols_to_drop = []

Nrows = df.shape[0]
# Go through each column except for the ones we want to drop
for col in df.columns.drop(cols_to_drop):
    # get unique values in column
    unique = df[col].unique();
    print("\n", col,":")
    # calculate frequency of each value
    for val in unique:
        if val not in ['nan']:
            N = len(df[df[col] == val])
            print("Value, N observations, frequency :", val," ,", N, " ,", np.round(N/Nrows,3))

    print("No. of NANs ", len(df[df[col].isna()]))



 Year Month :
Value, N observations, frequency : 201001.0  , 45  , 0.001
Value, N observations, frequency : 201002.0  , 46  , 0.001
Value, N observations, frequency : 201003.0  , 53  , 0.001
Value, N observations, frequency : 201004.0  , 55  , 0.001
Value, N observations, frequency : 201005.0  , 57  , 0.001
Value, N observations, frequency : 201006.0  , 62  , 0.001
Value, N observations, frequency : 201007.0  , 60  , 0.001
Value, N observations, frequency : 201008.0  , 64  , 0.001
Value, N observations, frequency : 201009.0  , 65  , 0.001
Value, N observations, frequency : 201010.0  , 65  , 0.001
Value, N observations, frequency : 201011.0  , 70  , 0.001
Value, N observations, frequency : 201012.0  , 72  , 0.001
Value, N observations, frequency : 201101.0  , 66  , 0.001
Value, N observations, frequency : 201102.0  , 70  , 0.001
Value, N observations, frequency : 201103.0  , 73  , 0.001
Value, N observations, frequency : 201104.0  , 67  , 0.001
Value, N observations, frequency : 201105

In [97]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59770 entries, 0 to 59769
Data columns (total 12 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Year Month           12166 non-null  float64
 1   Region               12166 non-null  object 
 2   Account Description  12166 non-null  object 
 3   Product              12166 non-null  object 
 4   Company              4786 non-null   object 
 5   Size                 12166 non-null  object 
 6   Size Numeric         12166 non-null  float64
 7   Strength             12166 non-null  object 
 8   WHO ATC 5 Code       12166 non-null  object 
 9   Volume               12166 non-null  float64
 10  Value                12166 non-null  object 
 11  PRIS                 4787 non-null   float64
dtypes: float64(4), object(8)
memory usage: 5.5+ MB


In [98]:
df_dropped = df.dropna()
df_dropped

Year Month       Region                       Account Description  \
0       201001.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
1       201001.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
2       201001.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
3       201002.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
4       201002.0  Hovedstaden  4060 Centralapoteket, Region Hovedstaden   
...          ...          ...                                       ...   
4782    201310.0  Midtjylland             4398 Viborg Sygehus, apoteket   
4783    201311.0  Midtjylland             4398 Viborg Sygehus, apoteket   
4784    201311.0  Midtjylland             4398 Viborg Sygehus, apoteket   
4785    201312.0  Midtjylland             4398 Viborg Sygehus, apoteket   
4786    201312.0  Midtjylland             4398 Viborg Sygehus, apoteket   

       Product     Company              Size  Size Numeric      Strength  \
0       Cimzia  UCB Nordic            2 stk.           2.0        200 mg   
1     Remicade         MSD  1 htgl. a 100 mg         100.0        100 mg   
2      Stelara     Janssen             1 stk           1.0  45 mg/0,5 ml   
3       Cimzia  UCB Nordic            2 stk.           2.0        200 mg   
4     Remicade         MSD  1 htgl. a 100 mg         100.0        100 mg   
...        ...         ...               ...           ...           ...   
4782  Remicade         MSD           1 htgl.           1.0        100 mg   
4783    Cimzia  UCB Nordic            2 stk.           2.0        200 mg   
4784  Remicade         MSD           1 htgl.           1.0        100 mg   
4785    Cimzia  UCB Nordic            2 stk.           2.0        200 mg   
4786  Remicade         MSD           1 htgl.           1.0        100 mg   

     WHO ATC 5 Code  Volume       Value      PRIS  
0           L04AB05    19.0    160820.1   8464.22  
1           L04AB02  2096.0   9236652.8   4406.80  
2           L04AC05    23.0   622716.26  27074.62  
3           L04AB05    11.0    88919.05   8083.55  
4           L04AB02  2036.0   8972244.8   4406.80  
...             ...     ...         ...       ...  
4782        L04AB02   472.0  2028009.36   4296.63  
4783        L04AB05    45.0    354665.7   7881.46  
4784        L04AB02   410.0   1761618.3   4296.63  
4785        L04AB05    52.0   409835.92   7881.46  
4786        L04AB02   292.0  1254615.96   4296.63  

[4786 rows x 12 columns]

OBS they might not need to be droped. Maybe one of the dataframes spell the brick differently, affecting the merge.

#### 2. Convert features with potential to be numeric into numeric features
from the investigation of the features we can see that a number of features can be converted from object data types into integers or floats.

In [99]:
to_numeric = [
    'blood clot in heart', 
    'absence', 
    'life_expectancy_years', 
    'longterm_disease_andel',
    'overweight_andel',
    't2d_patients',
    'uddannelse_lang_amount'
]
for col in to_numeric:
    df_dropped[col] = df_dropped[col].str.replace(',', '.', regex=True).astype(float)

KeyError: 'blood clot in heart'

In [19]:
df_dropped.info()

<class 'pandas.core.frame.DataFrame'>
Index: 76099 entries, 0 to 222794
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   brick_nr                76099 non-null  float64
 1   brick                   76099 non-null  object 
 2   blood clot in heart     76099 non-null  float64
 3   absence                 76099 non-null  float64
 4   life_expectancy_years   76099 non-null  float64
 5   longterm_disease_andel  76099 non-null  float64
 6   overweight_andel        76099 non-null  float64
 7   t2d_patients            76099 non-null  float64
 8   uddannelse_lang_amount  76099 non-null  float64
 9   physical_activity       76099 non-null  object 
 10  YEAR_MONTH              76099 non-null  float64
 11  SECTOR_CODE             76099 non-null  object 
 12  VOLUME                  76099 non-null  float64
 13  VALUE                   76099 non-null  float64
 14  WHO_ATC_CODE            76099 non-null  ob

Another interesting column is the column of YEAR_MONTH.

In [20]:
df_dropped['YEAR_MONTH'].unique()

array([202009., 202008., 202007., 202102., 202005., 202101., 202010.,
       202103., 202011., 202012., 202004., 202006.])

Looking at the column we can see that it combines both year and month into one as a float. Possible in the future it might be interesting to look at each year or each month. So lets split it and make it into two columns; one for year and one for month

In [21]:
# We start by making the column into a string to make it easier to slice
df_dropped['YEAR_MONTH'] = df_dropped['YEAR_MONTH'].astype(str)

# We then slice the year from the first 4 digits
df_dropped['year'] = df_dropped['YEAR_MONTH'].str[:4]  
# We then slice the year from the 4th digit to the 6th digit
df_dropped['month'] = df_dropped['YEAR_MONTH'].str[4:6]
df_dropped

C:\Users\nped\AppData\Local\Temp\ipykernel_24712\2133612125.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dropped['YEAR_MONTH'] = df_dropped['YEAR_MONTH'].astype(str)
C:\Users\nped\AppData\Local\Temp\ipykernel_24712\2133612125.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dropped['year'] = df_dropped['YEAR_MONTH'].str[:4]
C:\Users\nped\AppData\Local\Temp\ipykernel_24712\2133612125.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .l

brick_nr        brick  blood clot in heart   absence  \
0          308.0     Aabenraa             0.013750  6.002000   
1          308.0     Aabenraa             0.013750  6.002000   
2          308.0     Aabenraa             0.013750  6.002000   
3          308.0     Aabenraa             0.013750  6.002000   
4          308.0     Aabenraa             0.013750  6.002000   
...          ...          ...                  ...       ...   
222790     207.0  Vordingborg             0.013833  6.345222   
222791     207.0  Vordingborg             0.013833  6.345222   
222792     207.0  Vordingborg             0.013833  6.345222   
222793     207.0  Vordingborg             0.013833  6.345222   
222794     207.0  Vordingborg             0.013833  6.345222   

        life_expectancy_years  longterm_disease_andel  overweight_andel  \
0                   80.925500                0.358000          0.206333   
1                   80.925500                0.358000          0.206333   
2                   80.925500                0.358000          0.206333   
3                   80.925500                0.358000          0.206333   
4                   80.925500                0.358000          0.206333   
...                       ...                     ...               ...   
222790              79.865444                0.379111          0.231143   
222791              79.865444                0.379111          0.231143   
222792              79.865444                0.379111          0.231143   
222793              79.865444                0.379111          0.231143   
222794              79.865444                0.379111          0.231143   

        t2d_patients  uddannelse_lang_amount    physical_activity YEAR_MONTH  \
0        3936.207344            11845.819974  0,33866666666666667   202009.0   
1        3936.207344            11845.819974  0,33866666666666667   202008.0   
2        3936.207344            11845.819974  0,33866666666666667   202008.0   
3        3936.207344            11845.819974  0,33866666666666667   202008.0   
4        3936.207344            11845.819974  0,33866666666666667   202009.0   
...              ...                     ...                  ...        ...   
222790   3873.451601            12043.591696                0,334   202007.0   
222791   3873.451601            12043.591696                0,334   202007.0   
222792   3873.451601            12043.591696                0,334   202009.0   
222793   3873.451601            12043.591696                0,334   202010.0   
222794   3873.451601            12043.591696                0,334   202012.0   

       SECTOR_CODE  VOLUME    VALUE WHO_ATC_CODE  year month  
0                P    -1.0  -603.65      A10BH01  2020    09  
1                P     1.0   238.08      A10AC01  2020    08  
2                P    10.0  4894.70      A10AE06  2020    08  
3                P     3.0   891.87      A10AD01  2020    08  
4                P    25.0  6907.75      A10AC01  2020    09  
...            ...     ...      ...          ...   ...   ...  
222790           P     4.0  1266.88      A10BK03  2020    07  
222791           P     0.0   -53.38      A10BD08  2020    07  
222792           P   391.0  4301.00      A10BA02  2020    09  
222793           P     1.0   278.88      A10BD08  2020    10  
222794           P     3.0   981.30      A10BD20  2020    12  

[76099 rows x 17 columns]

#### 3. Another way to handle missing values
if we go back to the dataset from before we dropped nan values.

Here we can see that there are a number of missing values

In [22]:
df

brick_nr       brick blood clot in heart absence  \
0          308.0    Aabenraa             0,01375   6,002   
1          308.0    Aabenraa             0,01375   6,002   
2          308.0    Aabenraa             0,01375   6,002   
3          308.0    Aabenraa             0,01375   6,002   
4          308.0    Aabenraa             0,01375   6,002   
...          ...         ...                 ...     ...   
223696       NaN       Ærø                 NaN     NaN   
223697       NaN       Ærø                 NaN     NaN   
223698       NaN       Ærø                 NaN     NaN   
223699       NaN       Ærø                 NaN     NaN   
223700     403.0  Århus City               0,009    5,28   

       life_expectancy_years longterm_disease_andel     overweight_andel  \
0                    80,9255                  0,358  0,20633333333333334   
1                    80,9255                  0,358  0,20633333333333334   
2                    80,9255                  0,358  0,20633333333333334   
3                    80,9255                  0,358  0,20633333333333334   
4                    80,9255                  0,358  0,20633333333333334   
...                      ...                    ...                  ...   
223696                   NaN                    NaN                  NaN   
223697                   NaN                    NaN                  NaN   
223698                   NaN                    NaN                  NaN   
223699                   NaN                    NaN                  NaN   
223700                  81,5                  0,316                 0,11   

       t2d_patients uddannelse_lang_amount    physical_activity  YEAR_MONTH  \
0       3936,207344     11845,819973539024  0,33866666666666667    202009.0   
1       3936,207344     11845,819973539024  0,33866666666666667    202008.0   
2       3936,207344     11845,819973539024  0,33866666666666667    202008.0   
3       3936,207344     11845,819973539024  0,33866666666666667    202008.0   
4       3936,207344     11845,819973539024  0,33866666666666667    202009.0   
...             ...                    ...                  ...         ...   
223696          NaN                    NaN                  NaN    202006.0   
223697          NaN                    NaN                  NaN    202007.0   
223698          NaN                    NaN                  NaN    202008.0   
223699          NaN                    NaN                  NaN    202011.0   
223700   2169,52533     23503,461421150456                 0,24         NaN   

       SECTOR_CODE  VOLUME    VALUE WHO_ATC_CODE  
0                P    -1.0  -603.65      A10BH01  
1                P     1.0   238.08      A10AC01  
2                P    10.0  4894.70      A10AE06  
3                P     3.0   891.87      A10AD01  
4                P    25.0  6907.75      A10AC01  
...            ...     ...      ...          ...  
223696           P     2.0   708.98      A10AB05  
223697           P     3.0  3087.69      A10BK01  
223698           P    -3.0 -3145.80      A10BK01  
223699           P     1.0   622.85      A10BJ06  
223700         NaN     NaN      NaN          NaN  

[223701 rows x 15 columns]

In [23]:
# Let's look at the number of Nans for each numeric column
cols = df.select_dtypes(include=['float64', 'int64','object']).columns

Nrows = df.shape[0]
# Go through each column except for the ones we want to drop
for col in cols:
    # get unique values in column
    unique = df[col].unique();
    print(col,":")
    print("No. of NANs ", len(df[df[col].isna()]))



brick_nr :
No. of NANs  147574
brick :
No. of NANs  0
blood clot in heart :
No. of NANs  147574
absence :
No. of NANs  147574
life_expectancy_years :
No. of NANs  147574
longterm_disease_andel :
No. of NANs  147574
overweight_andel :
No. of NANs  147574
t2d_patients :
No. of NANs  147574
uddannelse_lang_amount :
No. of NANs  147574
physical_activity :
No. of NANs  147574
YEAR_MONTH :
No. of NANs  28
SECTOR_CODE :
No. of NANs  28
VOLUME :
No. of NANs  28
VALUE :
No. of NANs  28
WHO_ATC_CODE :
No. of NANs  28


Here we see that there are a number of missing values. Some of these can be handled through the merging between the dataframe SALES.csv and others.

Because of this we are not gonna focus on those values. But what about 'volume' and 'value'. Both columns are from the SALES dataset and are both contain float values. As such we are able to use a imputer to fill the missing value.

This method allows us to make a predictive model that predicts the missing values.

In [24]:
# Imputation functions
from sklearn.impute import KNNImputer, SimpleImputer

num_cols = ['VOLUME','VALUE']
# Too impute missing values, we introduce the following functions

# This imputer only simply replace the missing values with mean, median or most frequent value of that feature
simple_imputer = SimpleImputer(missing_values=np.nan, strategy='mean') 

# This imputer considers all numeric features and assigns to each missing value the feature mean of the n nearest neighbors
knn_imputer = KNNImputer(missing_values = np.nan, n_neighbors = 5)

# We will use the knn imputer, since it considers all numeric features to find the most similar neighbors:
df.loc[:, num_cols] = knn_imputer.fit_transform(df.loc[:, num_cols] )
df

brick_nr       brick blood clot in heart absence  \
0          308.0    Aabenraa             0,01375   6,002   
1          308.0    Aabenraa             0,01375   6,002   
2          308.0    Aabenraa             0,01375   6,002   
3          308.0    Aabenraa             0,01375   6,002   
4          308.0    Aabenraa             0,01375   6,002   
...          ...         ...                 ...     ...   
223696       NaN       Ærø                 NaN     NaN   
223697       NaN       Ærø                 NaN     NaN   
223698       NaN       Ærø                 NaN     NaN   
223699       NaN       Ærø                 NaN     NaN   
223700     403.0  Århus City               0,009    5,28   

       life_expectancy_years longterm_disease_andel     overweight_andel  \
0                    80,9255                  0,358  0,20633333333333334   
1                    80,9255                  0,358  0,20633333333333334   
2                    80,9255                  0,358  0,20633333333333334   
3                    80,9255                  0,358  0,20633333333333334   
4                    80,9255                  0,358  0,20633333333333334   
...                      ...                    ...                  ...   
223696                   NaN                    NaN                  NaN   
223697                   NaN                    NaN                  NaN   
223698                   NaN                    NaN                  NaN   
223699                   NaN                    NaN                  NaN   
223700                  81,5                  0,316                 0,11   

       t2d_patients uddannelse_lang_amount    physical_activity  YEAR_MONTH  \
0       3936,207344     11845,819973539024  0,33866666666666667    202009.0   
1       3936,207344     11845,819973539024  0,33866666666666667    202008.0   
2       3936,207344     11845,819973539024  0,33866666666666667    202008.0   
3       3936,207344     11845,819973539024  0,33866666666666667    202008.0   
4       3936,207344     11845,819973539024  0,33866666666666667    202009.0   
...             ...                    ...                  ...         ...   
223696          NaN                    NaN                  NaN    202006.0   
223697          NaN                    NaN                  NaN    202007.0   
223698          NaN                    NaN                  NaN    202008.0   
223699          NaN                    NaN                  NaN    202011.0   
223700   2169,52533     23503,461421150456                 0,24         NaN   

       SECTOR_CODE     VOLUME        VALUE WHO_ATC_CODE  
0                P  -1.000000  -603.650000      A10BH01  
1                P   1.000000   238.080000      A10AC01  
2                P  10.000000  4894.700000      A10AE06  
3                P   3.000000   891.870000      A10AD01  
4                P  25.000000  6907.750000      A10AC01  
...            ...        ...          ...          ...  
223696           P   2.000000   708.980000      A10AB05  
223697           P   3.000000  3087.690000      A10BK01  
223698           P  -3.000000 -3145.800000      A10BK01  
223699           P   1.000000   622.850000      A10BJ06  
223700         NaN  16.634623  4946.722091          NaN  

[223701 rows x 15 columns]

Here an example is shown that a prediction is made of "16.63" for VOLUME, and "4946.722" for VALUE

#### 4. Convert categorical features to numeric ones
- Convert features with two values to boolean features
- Label encode features with order
- One hot encoding

In our dataset we have the column SECTOR_CODE with *P* and *S*
this represent XXX.


In [25]:
df_dropped['SECTOR_CODE'].unique()

array(['P', 'S'], dtype=object)

If we wish to utilise this function for predictive modelling then we would have to change it into a numeric representation.
As there are only two different values it would be perfect to convert it into a boolean (0,1)

In [26]:
# First we describe the numeric representation
sector_code_rep = {'P': 0, 'S': 1}
df_dropped['SECTOR_CODE'] = df_dropped['SECTOR_CODE'].map(sector_code_rep)
df_dropped['SECTOR_CODE'].unique()

C:\Users\nped\AppData\Local\Temp\ipykernel_24712\1042394591.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dropped['SECTOR_CODE'] = df_dropped['SECTOR_CODE'].map(sector_code_rep)


array([0, 1], dtype=int64)

In [27]:
df_dropped

brick_nr        brick  blood clot in heart   absence  \
0          308.0     Aabenraa             0.013750  6.002000   
1          308.0     Aabenraa             0.013750  6.002000   
2          308.0     Aabenraa             0.013750  6.002000   
3          308.0     Aabenraa             0.013750  6.002000   
4          308.0     Aabenraa             0.013750  6.002000   
...          ...          ...                  ...       ...   
222790     207.0  Vordingborg             0.013833  6.345222   
222791     207.0  Vordingborg             0.013833  6.345222   
222792     207.0  Vordingborg             0.013833  6.345222   
222793     207.0  Vordingborg             0.013833  6.345222   
222794     207.0  Vordingborg             0.013833  6.345222   

        life_expectancy_years  longterm_disease_andel  overweight_andel  \
0                   80.925500                0.358000          0.206333   
1                   80.925500                0.358000          0.206333   
2                   80.925500                0.358000          0.206333   
3                   80.925500                0.358000          0.206333   
4                   80.925500                0.358000          0.206333   
...                       ...                     ...               ...   
222790              79.865444                0.379111          0.231143   
222791              79.865444                0.379111          0.231143   
222792              79.865444                0.379111          0.231143   
222793              79.865444                0.379111          0.231143   
222794              79.865444                0.379111          0.231143   

        t2d_patients  uddannelse_lang_amount    physical_activity YEAR_MONTH  \
0        3936.207344            11845.819974  0,33866666666666667   202009.0   
1        3936.207344            11845.819974  0,33866666666666667   202008.0   
2        3936.207344            11845.819974  0,33866666666666667   202008.0   
3        3936.207344            11845.819974  0,33866666666666667   202008.0   
4        3936.207344            11845.819974  0,33866666666666667   202009.0   
...              ...                     ...                  ...        ...   
222790   3873.451601            12043.591696                0,334   202007.0   
222791   3873.451601            12043.591696                0,334   202007.0   
222792   3873.451601            12043.591696                0,334   202009.0   
222793   3873.451601            12043.591696                0,334   202010.0   
222794   3873.451601            12043.591696                0,334   202012.0   

        SECTOR_CODE  VOLUME    VALUE WHO_ATC_CODE  year month  
0                 0    -1.0  -603.65      A10BH01  2020    09  
1                 0     1.0   238.08      A10AC01  2020    08  
2                 0    10.0  4894.70      A10AE06  2020    08  
3                 0     3.0   891.87      A10AD01  2020    08  
4                 0    25.0  6907.75      A10AC01  2020    09  
...             ...     ...      ...          ...   ...   ...  
222790            0     4.0  1266.88      A10BK03  2020    07  
222791            0     0.0   -53.38      A10BD08  2020    07  
222792            0   391.0  4301.00      A10BA02  2020    09  
222793            0     1.0   278.88      A10BD08  2020    10  
222794            0     3.0   981.30      A10BD20  2020    12  

[76099 rows x 17 columns]

#### Ranking ordered features (ordinal features) using label encoding:

Since one-hot encoding can increase the no. of features dramatically,
we will show another method of changing categorical features into numeric - LabelEncoding. Import to note here is that if we just use LabelEncoding then the categorical feature will be label in order of frequency. However, when we utilise this feature for training machine learning models, it will count the ones with higher values as being more important. This is not necessarily the truth. We wish for a way so that the order of numbers is representative. Thus we need to make an order to base our labeling on. Sometimes, the order is very clear (e.g. low, mid, high, very high), and sometimes we have to be
a bit creative.


In the following, we'll convert the brick and brick_nr by finding a metric for the order and then using label encoding. Looking at the dataframe there are many different features that could be the basis for creating the order. Just as an example we will use the number of type 2 diabetes (t2d_patients)

In [28]:
from sklearn.preprocessing import LabelEncoder

# Sort the DataFrame based on column_y to define the order
encoded_df = df_dropped.sort_values(by='longterm_disease_andel')

# Initialize the LabelEncoder
label_encoder = LabelEncoder()

# Fit the LabelEncoder on the sorted column_x and transform it
encoded_df['brick'] = label_encoder.fit_transform(encoded_df['brick'])

df_dropped['brick'] = encoded_df.set_index(df_dropped.index)['brick']
df_dropped

C:\Users\nped\AppData\Local\Temp\ipykernel_24712\1564712222.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dropped['brick'] = encoded_df.set_index(df_dropped.index)['brick']


brick_nr  brick  blood clot in heart   absence  life_expectancy_years  \
0          308.0      5             0.013750  6.002000              80.925500   
1          308.0      5             0.013750  6.002000              80.925500   
2          308.0      5             0.013750  6.002000              80.925500   
3          308.0      5             0.013750  6.002000              80.925500   
4          308.0      5             0.013750  6.002000              80.925500   
...          ...    ...                  ...       ...                    ...   
222790     207.0     17             0.013833  6.345222              79.865444   
222791     207.0     17             0.013833  6.345222              79.865444   
222792     207.0     17             0.013833  6.345222              79.865444   
222793     207.0     17             0.013833  6.345222              79.865444   
222794     207.0     17             0.013833  6.345222              79.865444   

        longterm_disease_andel  overweight_andel  t2d_patients  \
0                     0.358000          0.206333   3936.207344   
1                     0.358000          0.206333   3936.207344   
2                     0.358000          0.206333   3936.207344   
3                     0.358000          0.206333   3936.207344   
4                     0.358000          0.206333   3936.207344   
...                        ...               ...           ...   
222790                0.379111          0.231143   3873.451601   
222791                0.379111          0.231143   3873.451601   
222792                0.379111          0.231143   3873.451601   
222793                0.379111          0.231143   3873.451601   
222794                0.379111          0.231143   3873.451601   

        uddannelse_lang_amount    physical_activity YEAR_MONTH  SECTOR_CODE  \
0                 11845.819974  0,33866666666666667   202009.0            0   
1                 11845.819974  0,33866666666666667   202008.0            0   
2                 11845.819974  0,33866666666666667   202008.0            0   
3                 11845.819974  0,33866666666666667   202008.0            0   
4                 11845.819974  0,33866666666666667   202009.0            0   
...                        ...                  ...        ...          ...   
222790            12043.591696                0,334   202007.0            0   
222791            12043.591696                0,334   202007.0            0   
222792            12043.591696                0,334   202009.0            0   
222793            12043.591696                0,334   202010.0            0   
222794            12043.591696                0,334   202012.0            0   

        VOLUME    VALUE WHO_ATC_CODE  year month  
0         -1.0  -603.65      A10BH01  2020    09  
1          1.0   238.08      A10AC01  2020    08  
2         10.0  4894.70      A10AE06  2020    08  
3          3.0   891.87      A10AD01  2020    08  
4         25.0  6907.75      A10AC01  2020    09  
...        ...      ...          ...   ...   ...  
222790     4.0  1266.88      A10BK03  2020    07  
222791     0.0   -53.38      A10BD08  2020    07  
222792   391.0  4301.00      A10BA02  2020    09  
222793     1.0   278.88      A10BD08  2020    10  
222794     3.0   981.30      A10BD20  2020    12  

[76099 rows x 17 columns]

#### One-hot encoding

One-hot encoding has the advantage that is does not introduce any order between features like label encoding does, but the drawback that K different categories of a feature will be transformed into K or K-1 new features, and so the total number of features can become very large, which can hurt performance. For this reasons, if a feature has many categories, these categories are often grouped in straightforward (simply group rare categories) or clever (use domain knowledge to make groups)

We will one-hot encode the column 'WHO_ATC_CODE'
ATC: The Anatomical Therapeutic Chemical code - a code assigned to medicine based on it organic structure and/or system

In [29]:
df_dropped['WHO_ATC_CODE'].nunique()

47

In [30]:
## We need this function to do one-hot (or dummy) encoding
def do_dummy_encoding(dataframe, target_columns, prefix_list, drop_first = False):
    """
    For each cateogrical column with categories (l1,...,ln), this function transform the category into n ( if drop_first = False)
    or n-1 columns, where [0,0,...,0] represents l1,
    [1,0,...,0] rep. l2, [0,1,0,...,0] rep l3 etc
    """
    data = dataframe.copy()
    # Build dummy index dataframe
    dummy_columns = pd.get_dummies(data[target_columns], prefix = prefix_list, drop_first=drop_first)

    ## Append new columns to dataframe
    data = pd.concat([data, dummy_columns], axis = 1)

    ## Drop original columns
    data.drop(columns = target_columns, inplace=True)
    return data

In [31]:
onehot_cols = ['WHO_ATC_CODE']

for i, col in enumerate(onehot_cols):
    # Use onehot_cols[i] as prefix for the new columns
    # If drop_first = False, the first category will be included in the new columns
    # If drop_first = True, the first category will be dropped, meaning that
    #  if all new columns are 0, the category is the first one. This is the most
    #  efficient way of doing it, but it has the price that we have to remember that
    # the first category is represented like this
    df_dropped = do_dummy_encoding(df_dropped, col, onehot_cols[i], drop_first = False)


In [32]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 223701 entries, 0 to 223700
Data columns (total 15 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   brick_nr                76127 non-null   float64
 1   brick                   223701 non-null  object 
 2   blood clot in heart     76127 non-null   object 
 3   absence                 76127 non-null   object 
 4   life_expectancy_years   76127 non-null   object 
 5   longterm_disease_andel  76127 non-null   object 
 6   overweight_andel        76127 non-null   object 
 7   t2d_patients            76127 non-null   object 
 8   uddannelse_lang_amount  76127 non-null   object 
 9   physical_activity       76127 non-null   object 
 10  YEAR_MONTH              223673 non-null  float64
 11  SECTOR_CODE             223673 non-null  object 
 12  VOLUME                  223701 non-null  float64
 13  VALUE                   223701 non-null  float64
 14  WHO_ATC_CODE        

In [35]:
df_dropped

brick_nr  brick  blood clot in heart   absence  life_expectancy_years  \
0          308.0      5             0.013750  6.002000              80.925500   
1          308.0      5             0.013750  6.002000              80.925500   
2          308.0      5             0.013750  6.002000              80.925500   
3          308.0      5             0.013750  6.002000              80.925500   
4          308.0      5             0.013750  6.002000              80.925500   
...          ...    ...                  ...       ...                    ...   
222790     207.0     17             0.013833  6.345222              79.865444   
222791     207.0     17             0.013833  6.345222              79.865444   
222792     207.0     17             0.013833  6.345222              79.865444   
222793     207.0     17             0.013833  6.345222              79.865444   
222794     207.0     17             0.013833  6.345222              79.865444   

        longterm_disease_andel  overweight_andel  t2d_patients  \
0                     0.358000          0.206333   3936.207344   
1                     0.358000          0.206333   3936.207344   
2                     0.358000          0.206333   3936.207344   
3                     0.358000          0.206333   3936.207344   
4                     0.358000          0.206333   3936.207344   
...                        ...               ...           ...   
222790                0.379111          0.231143   3873.451601   
222791                0.379111          0.231143   3873.451601   
222792                0.379111          0.231143   3873.451601   
222793                0.379111          0.231143   3873.451601   
222794                0.379111          0.231143   3873.451601   

        uddannelse_lang_amount    physical_activity  ... WHO_ATC_CODE_A10BJ01  \
0                 11845.819974  0,33866666666666667  ...                False   
1                 11845.819974  0,33866666666666667  ...                False   
2                 11845.819974  0,33866666666666667  ...                False   
3                 11845.819974  0,33866666666666667  ...                False   
4                 11845.819974  0,33866666666666667  ...                False   
...                        ...                  ...  ...                  ...   
222790            12043.591696                0,334  ...                False   
222791            12043.591696                0,334  ...                False   
222792            12043.591696                0,334  ...                False   
222793            12043.591696                0,334  ...                False   
222794            12043.591696                0,334  ...                False   

        WHO_ATC_CODE_A10BJ02  WHO_ATC_CODE_A10BJ03  WHO_ATC_CODE_A10BJ05  \
0                      False                 False                 False   
1                      False                 False                 False   
2                      False                 False                 False   
3                      False                 False                 False   
4                      False                 False                 False   
...                      ...                   ...                   ...   
222790                 False                 False                 False   
222791                 False                 False                 False   
222792                 False                 False                 False   
222793                 False                 False                 False   
222794                 False                 False                 False   

       WHO_ATC_CODE_A10BJ06 WHO_ATC_CODE_A10BK01  WHO_ATC_CODE_A10BK02  \
0                     False                False                 False   
1                     False                False                 False   
2                     False                False                 False   
3                     False                False            

In [3]:
diagnose_grupper = pd.read_csv('Tabel_SKS_Diagnose_hospital.csv', sep=";")
diagnose_grupper.info()

NameError: name 'pd' is not defined